packages/contracts-bedrock/src/L2/GasTank.sol

```diff
@@ -0,0 +1,119 @@
+
+        // Record the pending withdrawal
+        withdrawals[msg.sender] = Withdrawal({ timestamp: block.timestamp, amount: amount });
+
+        // Emit an event for the withdrawal initiation
+        emit WithdrawalInitiated(msg.sender, amount);
+    }
+
+    function finalizeWithdrawal(address to) external {
+        Withdrawal memory withdrawal = withdrawals[msg.sender];
+
+        // Ensure the withdraw is not pending
+        if (block.timestamp < withdrawal.timestamp + WITHDRAWAL_DELAY) revert WithdrawPending();
+
+        // Update the balance
+        uint256 amount = balanceOf[msg.sender] < withdrawal.amount ? balanceOf[msg.sender] : withdrawal.amount;
+        balanceOf[msg.sender] -= amount;
+
+        // Clear the pending withdrawal
+        delete withdrawals[msg.sender];
+
+        // Send the funds to the recipient
+        new SafeSend{ value: amount }(payable(to));
+
+        emit WithdrawalFinalized(msg.sender, to, amount);
+    }
+
+    // Flag a message into the gas tank so the relayer is aware of it, and can claim the funds after relaying
+    function flag(bytes32 rootMessageHash) external {
+        flaggedMessages[msg.sender][rootMessageHash] = true;
+        emit Flagged(rootMessageHash, msg.sender);
+    }
+
+    // Claim repayment for a relayed message
+    function claim(Identifier calldata id, address gasProvider, bytes calldata payload) external {
+        // Ensure the origin is the messenger
+        if (id.origin != address(MESSENGER)) revert InvalidOrigin();
+
+        // Decode the receipt
+        if (bytes32(payload[:32]) != RelayedMessageGasReceipt.selector) revert InvalidPayload();
+        (bytes32 msgHash, bytes32 rootMsgHash, address relayer, uint256 relayCost) = decodeGasReceiptPayload(payload);
+
+        // Ensure the message is flagged for relaying
+        if (!flaggedMessages[gasProvider][rootMsgHash]) revert InvalidPayer();
+
+        // Ensure unclaimed
+        if (claimed[msgHash]) revert AlreadyClaimed();
+
+        // Compute total cost (adding the overhead of this claim)
+        uint256 claimCost = CLAIM_OVERHEAD * block.basefee;
+        uint256 cost = relayCost + claimCost;
+        // TODO: Make it more flexible so to allow partial repayment, but tracking the claim as partially repaid
+        if (balanceOf[gasProvider] < cost) revert InsufficientBalance();
+
+        // Ensure the original outbound message was sent from this chain
+        if (!MESSENGER.sentMessages(rootMsgHash)) revert InvalidRootMessage();
+
+        // Validate the message
+        ICrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(id, keccak256(payload));
+
+        // Update the balance and mark the claim
+        balanceOf[gasProvider] -= cost;
+        claimed[msgHash] = true;
+
+        // Send the cost repayment back to the relayer
+        new SafeSend{ value: cost }(payable(relayer));
+
+        emit Claimed(msgHash, relayer, cost);
```

Comment:
can you also emit the `gasProvider` and the `rootMsgHash`? this will help with enabling the relayer to calculate `balanceOf[gasProvider]` and to also determine how many claims have occurred for a given `rootMessageHash`.